src/hyperdeck/response-handler.js

```diff
@@ -12,73 +12,78 @@ var ASYNC_RESPONSE_REGEX = /^(5[0-9]{2} )/;
 function ResponseHandler(clientSocket) {
   var destroyed = false;
   var notifier = new events.EventEmitter();
-  var buffer = "";
-  var nextResponseIsMultiline = false;
-  
-  function isBufferComplete() {
-    var lines = buffer.split("\r\n");
-    // there will always be an empty element at the end as every line will always end in "\r\n"
-    // so remove it.
-    // i.e. "1\r\n2\r\n" => ["1", "2", ""]
-    lines.pop();
-    if (lines.length === 1) {
-      var asyncResponse = ASYNC_RESPONSE_REGEX.exec(lines[0]);
-      if (!asyncResponse) {
-        if (nextResponseIsMultiline) {
-          logger.debug('Assuming buffer is not complete because nextResponseIsMultiline flag is set.');
-          nextResponseIsMultiline = false;
-          return false;
-        }
-      }
-      // is it a single line response?
-      return lines[0].indexOf(":") !== lines[0].length-1;
-    } 
-    // multi line response so waiting for a blank line to signify end
-    return lines[lines.length-1] === "";
+  var buffer = [];
+
+  function isRespComplete() {
+    let complete = false;
+
+    if (buffer.length === 1) {
+      // a single line response
+      // single line response first line should never ends with ":" and
+      // 1XX and 200 is always single line response
+      complete = buffer[0].search(/^1\d\d\s|200\sok$/) === 0 && buffer[0].endsWith(':') === false;
+    } else if (buffer.length >= 3) {
+      // multi line response, so waiting for a blank line to signify end
+      complete = buffer[buffer.length - 1] === '';
+    }
+
+    return complete;
   }
-  
+
   function onData(rawData) {
-    logger.debug('Got data on socket.', rawData);
-    rawData.split("\r\n").forEach(function(line) {
-      if (buffer === '' && line.trim() === '') {
-        // handle empty lines before the data
-        // see https://github.com/LA1TV/Hyperdeck-JS-Lib/issues/44
-        return;
-      }
-      buffer += (line + "\r\n");
-      if (!isBufferComplete()) {
-        return;
+    logger.debug('Got data on socket.\n', rawData);
+
+    const resArray = rawData.split('\r\n');
+    // every response(even one line) must always end with \r\n, if not its not a valid response
+    if (resArray.length === 0 || resArray[resArray.length - 1] != '') {
+      throw "Unknown response type.";
+    }
+
+    resArray.pop();
+    resArray.forEach(function (line) {
+      // push to buffer till response is read completly
+      if (buffer.length > 0 || (buffer.length === 0 && line)) {
```

Comment:
I think it's safer to do `line.trim() === ''` and it would be good to keep the comment :)
```
// handle empty lines before the data
// see https://github.com/LA1TV/Hyperdeck-JS-Lib/issues/44
```

Reply:
in [ help ] command response some lines are "\s" (line with one space), so its better not to use line.trim() === ''

Reply:
But this only applies when we have nothing in the buffer yet, so waiting for the first line

Reply:
correct, in my mind, your code was hovering ;)